### PR TITLE
feat: support badge number on web

### PR DESCRIPTION
**Pull request by jaulz**
# Why

Badge numbers were not supported on web so far. 

# How

This solution uses https://www.npmjs.com/package/@approvals-cloud/badgin which is a wrapper around the future Badging API that nicely falls back to the favicon or title.

# Test Plan

Use `getBadgeNumberAsync` and `setBadgeNumberAsync` on web.

### CODE REVIEW

File: packages/expo/src/Notifications/Notifications.ts [hunk 1]
```diff
@@ -418,16 +418,10 @@ export default {
...
 
   async getBadgeNumberAsync(): Promise<number> {
-    if (!ExponentNotifications.getBadgeNumberAsync) {
-      return 0;
-    }
     return ExponentNotifications.getBadgeNumberAsync();
   },
 
   async setBadgeNumberAsync(number: number): Promise<void> {
-    if (!ExponentNotifications.setBadgeNumberAsync) {
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/Notifications.ts]: let's leave these in for supporting other platforms.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 2]
```diff
@@ -18,6 +20,19 @@ export default {
...
+    return currentBadgeNumber;
+  },
+  async setBadgeNumberAsync(badgeNumber: number): Promise<void> {
+    currentBadgeNumber = badgeNumber;
+    
+    // Update title
+    const match = document.title.match(new RegExp('\(\d*\) (.*)'));
+    const originalTitle = match ? match[2] : document.title;
+    if (currentBadgeNumber) {
+      document.title = `(${currentBadgeNumber}) ${originalTitle}`;
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: We may need to observe the document title as react-navigation updates it.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 3]
```diff
@@ -18,6 +20,19 @@ export default {
   async dismissAllNotifications(): Promise<void> {},
   async cancelScheduledNotification(notificationId: string): Promise<void> {},
   async cancelAllScheduledNotifications(): Promise<void> {},
-  //   getBadgeNumberAsync(): Promise<void> {},
-  //   setBadgeNumberAsync(badgeNumber: number): Promise<void> {},
+  async getBadgeNumberAsync(): Promise<number> {
+    return currentBadgeNumber;
+  },
+  async setBadgeNumberAsync(badgeNumber: number): Promise<void> {
+    currentBadgeNumber = badgeNumber;
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: Possibly write this to `AsyncStorage`.

**Review by EvanBacon — CHANGES_REQUESTED** [review]

File: packages/expo/src/Notifications/Notifications.ts (hunk 1 above)
Comment (jaulz) [reply packages/expo/src/Notifications/Notifications.ts]: Okay, it's restored.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 3 above)
Comment (jaulz) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: I think we can keep it as a local variable as it does not need to retrieved somewhere else. That would also keep it as simple as possible.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 2 above)
Comment (jaulz) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: I created a package https://www.npmjs.com/package/@approvals-cloud/badgin that uses the favicon first but if no favicon is available it proxies the title and also covers changes of the title.

File: packages/expo/package.json [hunk 4]
```diff
@@ -55,6 +55,7 @@
   },
   "homepage": "https://github.com/expo/expo/tree/master/packages/expo",
   "dependencies": {
+    "@approvals-cloud/badgin": "^1.1.0",
```
Comment (sjchmiela) [inline packages/expo/package.json]: Would you care to change it to `badgin`?

Zrzut ekranu 2019-11-20 o 17 31 51

**Review by sjchmiela — CHANGES_REQUESTED** [review]
Apart from that one comment looks good to me!

**Comment by sjchmiela** [comment]
Could you please also add the `CHANGELOG` entry for this feature? 🙂

**Comment by jaulz** [comment]
Sure, will do it by end of next week since I am currently on vacation :blush:

**Comment by jaulz** [comment]
@sjchmiela I just implemented your suggestions. Anything else? 😊

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 5]
```diff
@@ -110,4 +113,12 @@ export default {
   async getDevicePushTokenAsync(): Promise<{ type: string; data: Object }> {
     return await getDevicePushTokenAsync();
   },
+  
+  async getBadgeNumberAsync(): Promise<number> {
+    return currentBadgeNumber;
+  },
+  async setBadgeNumberAsync(badgeNumber: number): Promise<void> {
+    currentBadgeNumber = badgeNumber
+    badgin.set(badgeNumber)
```
Comment (ide) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: Does this clear when set to zero?

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 6]
```diff
@@ -110,4 +113,12 @@ export default {
   async getDevicePushTokenAsync(): Promise<{ type: string; data: Object }> {
     return await getDevicePushTokenAsync();
   },
+  
+  async getBadgeNumberAsync(): Promise<number> {
+    return currentBadgeNumber;
+  },
```
Comment (ide) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]:
```suggestion
  },

```

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 7]
```diff
@@ -110,4 +113,12 @@ export default {
...
+  async setBadgeNumberAsync(badgeNumber: number): Promise<void> {
+    currentBadgeNumber = badgeNumber
```
Comment (ide) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]:
```suggestion
    currentBadgeNumber = badgeNumber;
```

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 5 above)
Comment (ide) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]:
```suggestion
    badgin.set(badgeNumber);
```

File: packages/expo/package.json [hunk 8]
```diff
@@ -58,6 +58,7 @@
     "@types/uuid-js": "^0.7.1",
     "@unimodules/core": "~5.0.0",
     "@unimodules/react-native-adapter": "~5.0.0",
+    "badgin": "^1.0.1",
```
Comment (ide) [inline packages/expo/package.json]: Need to update the lockfile

**Review by ide — CHANGES_REQUESTED** [review]
This looks OK to try for now but since a lot of the functionality is in the package if there are issues we might need to revert.

File: packages/expo/package.json (hunk 8 above)
Comment (jaulz) [reply packages/expo/package.json]: It's updated now.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 5 above)
Comment (jaulz) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: Yes, it does clear the badge if set to zero. I just released a new version that explicitly checks this case.

File: yarn.lock [hunk 9]
```diff
@@ -13240,6 +13245,30 @@ react-native-tab-view@^2.9.0:
   resolved "https://registry.yarnpkg.com/react-native-tab-view/-/react-native-tab-view-2.10.0.tgz#5e249e5650502010013449ffd4e5edc18a95364b"
   integrity sha512-qgexVz5eO4yaFjdkmn/sURXgVvaBo6pZD/q1eoca96SbPVbaH3WzVhF3bRUfeTHwZkXwznFTpS3JURqIFU8vQA==
 
+react-native-unimodules@~0.7.0, react-native-unimodules@~0.7.0-rc.3:
```
Comment (ide) [inline yarn.lock]: Revert, you need to check out submodules first.

Comment (jaulz) [reply yarn.lock]: Hm, better now?

**Comment by ide** [comment]
Looks good, just one more CI test that needs to pass:

```
🔍 Checking the expo package ...
🏃‍♀️ Running `yarn clean` ...
🏃‍♀️ Running `yarn build` ...
The following build files need to be rebuilt and committed:
build/Notifications/ExponentNotifications.web.d.ts
build/Notifications/ExponentNotifications.web.js
build/Notifications/ExponentNotifications.web.js.map
```

**Comment by jaulz** [comment]
Ah, I didn't know that the build files also need to be committed. Checking all the test results I think the PR is finally ready. As far as I could see the failed tests are not related to these changes.

**Pull request merged**